### PR TITLE
Add flex

### PR DESCRIPTION
**Pull request by jakirkham**
Replaces https://github.com/conda-forge/staged-recipes/pull/120

Builds flex for Linux and Mac. Borrowed with some modifications from conda recipes.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @ocefpaf

**Comment by jakirkham** [comment]
Yeah, this is already behaving better. Circle CI was able to get the commit this time. Travis CI started. AppVeyor is skipped. Not sure what happened with the old one, but this is working.

**Comment by jakirkham** [comment]
`make check` failing on Circle CI will investigate.

**Comment by groutr** [comment]
Any progress on this?

**Comment by jakirkham** [comment]
Sorry, haven't had time to dig into this of late. If you want to try fixing it, feel free to PR changes against this and we can get it across the finish line.

**Comment by groutr** [comment]
It's possible that `make check` won't work since there is a circular dependency between flex and bison.  The test suite for flex requires bison, but bison requires flex to build.

**Comment by jakirkham** [comment]
Ah, ok. Good point, @groutr. :100: Thanks for getting my head out of the sand.

Will comment `make check` here and we can re-enable after bison is built and added.

**Comment by jakirkham** [comment]
@ocefpaf, could you please give this a review? It is one of my older recipes so may still have some lingering issues. I tried to fix it up as best I could.

File: recipes/flex/build.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+./configure --prefix="$PREFIX"
+make
+# Skipped as this requires bison.
+# Bison requires flex to build.
+# So, will have to do the first
+#  round of builds without this.
+#make check
```
Comment (ocefpaf) [inline recipes/flex/build.sh]: :+1:

**Comment by jakirkham** [comment]
NVM. Had to upgrade `flex` to fix a Mac bug. Now we need `yacc` to build it.

File: recipes/flex/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+test:
+  commands:
+    # CLI tests
+    - flex --help
+
+    # Include test
+    - test -f ${PREFIX}/include/FlexLexer.h
+
+    # Lib tests
```
Comment (ocefpaf) [inline recipes/flex/meta.yaml]: I am not a big fan of these spaces, but that is not important.

File: recipes/flex/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+source:
+  fn: flex-{{ version }}.tar.gz
+  url: http://sourceforge.net/projects/flex/files/flex-{{ version }}.tar.gz
+  md5: 5724bcffed4ebe39e9b55a9be80859ec
+
+build:
+  skip: true  # [win]
+  number: 0
+  detect_binary_files_with_prefix: true
```
Comment (ocefpaf) [inline recipes/flex/meta.yaml]: In theory doesn't that the defaults to `True`?

**Comment by ocefpaf** [comment]
LGTM, but I guess you'll need `yacc`. Hopefully that is not another circular dep.

```shell
../build-aux/ylwrap: line 176: yacc: command not found
make[1]: *** [bison_nr_parser.c] Error 127
make[1]: Leaving directory `/opt/conda/conda-bld/work/flex-2.6.0/tests'
make: *** [all-recursive] Error 1
```

File: recipes/flex/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/flex/meta.yaml]: Actually the [opposite]( http://conda.pydata.org/docs/building/meta-yaml.html#detect-binary-files-with-prefix ). Though I wonder if this isn't a hold over from earlier conda build days.

**Comment by jakirkham** [comment]
Good point. As `bison` subsumes `yacc` IIRC, let's see if I can't apply the patch for Mac so we can at least get one of these things done.

**Comment by jakirkham** [comment]
Alright, I tried adding the patch for Mac. Let's see how this goes.

**Comment by jakirkham** [comment]
If this passes, are we ok with merging as is or there any serious concerns? If there are serious issues, I should be able to look again tomorrow.

File: recipes/flex/meta.yaml (hunk 3 above)
Comment (pelson) [inline recipes/flex/meta.yaml]: We can always leave it out, and if it is a problem fix it + add a test :wink:

**Comment by groutr** [comment]
I've seen in the past is a small wrapper for `bison -y` to `yacc` (the `-y` is the yacc compatibility mode)

**Comment by jakirkham** [comment]
Yes, `bison` is designed so that it can be fully `yacc` compliant. Honestly, I'm a bit unsure of where one gets `yacc` source code and from my understanding there are a few variants. So, they may not all behave the same.

**Comment by jakirkham** [comment]
I should add that `bison` also includes a `yacc` binary with it.

**Pull request merged**